ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -140,20 +145,103 @@ class GetBlobOperation extends GetOperation {
     super(routerConfig, routerMetrics, clusterMap, responseHandler, blobId, options, callback, kms, cryptoService,
         cryptoJobHandler, time, isEncrypted);
     this.routerCallback = routerCallback;
     this.blobIdFactory = blobIdFactory;
     this.quotaChargeCallback = quotaChargeCallback;
-    firstChunk = new FirstGetChunk();
+    this.blobMetadataCache = blobMetadataCache;
+    if (shouldLookupMetadataCache()) {
+      BlobMetadata blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+      if (blobMetadata != null) {
+        firstChunk = new CachedFirstChunk(blobMetadata);
+        blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheHit;
+      } else {
+        firstChunk = new FirstGetChunk();
+        blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheMiss;
+      }
+    } else {
+      firstChunk = new FirstGetChunk();
+    }
+  }
+
+  /**
+   * Decides if we should look up the metadata cache during a GET request
+   * @return True if and only if
+   *  - cache exists
+   *  - it is a RANGE request
+   *  - it is not a raw mode request
+   *  - the blobID is for a metadata blob
+   *  False, otherwise.
+   */
+  boolean shouldLookupMetadataCache() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
+        options.getBlobOptions.isRawMode() == false &&
+        blobId.getBlobDataType() == BlobId.BlobDataType.METADATA);
+  }
+
+  /**
+   * Decides if we should cache metadata during a GET request
+   * This method should be called after decryption is complete.
+   * @return True if and only if
+   *  - cache exists
+   *  - it is RANGE request
+   *  - it is not a raw mode request
+   *  - the blobID is for a metadata blob
+   *  - the cache was looked up and we got a cache miss
+   *  False, otherwise.
+   */
+  boolean shouldSaveMetadata() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
+        options.getBlobOptions.isRawMode() == false &&
+        blobId.getBlobDataType() == BlobId.BlobDataType.METADATA &&
+        totalSize >= routerConfig.routerSmallestBlobForMetadataCache &&
+        isMetadataCacheMiss());
+  }
+
+  /**
+   * Conditionally saves blob metadata for composite blobs
+   * @return True if metadata was saved successfully, else False.
+   */
+  boolean saveMetadata() {
+    boolean putResult = false;
+    if (shouldSaveMetadata()) {
+      BlobMetadata blobMetadata = new BlobMetadata(blobId.toString(), blobInfo, compositeBlobInfo);
+      putResult = blobMetadataCache.putObject(blobMetadata.getBlobId(), blobMetadata);
+      logger.trace("Saving metadata for blobId = {}, result = {}", blobId, putResult);
+    }
+    return putResult;
+  }
+
+  /**
+   * Unconditionally deletes blob metadata for composite blobs.
+   * If it was never saved, then we don't have to delete it.
+   * @return True if metadata was deleted successfully, else False.
+   * @param reason
```

Comment:
Good point. The return val from this method is not currently used but, in general, i think its a good practice to return something assuming it may be used in the future.